### PR TITLE
Use a Timings* in place of optional<Timings>*

### PR DESCRIPTION
**Pull request by danakj**
The current representation has two nested presence indicatators (the optional bool, the null pointer). Currently the pointer is never null, but the style guide suggests that T* should be used for parameters that may or may not be present, so we do not need the optional here.

> When passing an object's address as an argument, use a reference
> unless one of the following cases applies:
>
> - If the parameter is optional, use a pointer and document that it
>   may be null.

Once the parameter is just a pointer, the ScopedTiming field does not need an optional either, and can just store the pointer.

It would be more preferable to have an optional representation of a sometimes-null pointer like optional<T&> to describe a sometimes-null pointer, as this would allow clearer runtime diagnostics when used incorrectly (a check failure in unwrapping) and would be better self-documenting through syntax instead of a comment. But we do not currently have such a primitive.

### CODE REVIEW

File: toolchain/driver/compile_subcommand.cpp [hunk 1]
```diff
@@ -443,7 +443,7 @@ class CompilationUnit {
     sem_ir_converter_.emplace(node_converters, &*sem_ir_);
     return {.consumer = consumer_,
             .value_stores = &value_stores_,
-            .timings = &timings_,
+            .timings = timings_ ? &*timings_ : nullptr,
```
Comment (ZekeMiller) [inline toolchain/driver/compile_subcommand.cpp:446]: [abseil.io/tips/141](https://abseil.io/tips/141) makes a case in favor of explicit `.has_value()` for clarity

(I'm a new contributor, LMK if there's already a style decision on this!)

Comment (danakj) [reply toolchain/driver/compile_subcommand.cpp:446]: I believe that is more for `optional<T>` where `T` converts to `bool`, as it becomes more ambiguous.

File: toolchain/base/timings.h [hunk 2]
```diff
@@ -14,24 +16,24 @@ namespace Carbon {
...
+      if (timings) {
+        timings->Add(label, std::chrono::steady_clock::now() - start);
       }
     }
 
    private:
-    std::optional<Timings>* timings;
+    Timings* timings;
     llvm::StringRef label;
     std::chrono::steady_clock::time_point start;
```
Comment (jonmeow) [inline toolchain/base/timings.h]: Small thing while you're in here, I should've had these class members named with underscores (`timings_` etc). Would you be okay fixing that as long as you're refactoring here? (if you'd rather not, feel free to merge without it)

File: toolchain/driver/compile_subcommand.cpp (hunk 1 above)
Comment (jonmeow) [reply toolchain/driver/compile_subcommand.cpp:446]: FWIW I think we don't generally use that tip. i.e., we don't often write `!= nullptr` or `.has_value()`. We do generally avoid implicit condition checks for ints (`!= 0`) and enums.

i.e., as-is looks good.

**Review by jonmeow — APPROVED** [review]
LG, the way this ends up being stored on `Unit` feels more consistent.

To be sure, the pointer style bit is more intended for whether to pass `Timings*` or `Timings&` when storing a reference, rather than `std::optional`. I'd been going back and forth about how to pass `timings` because I'm mixed about the presence checks in `compile_subcommand.cpp`, but probably this is fine.

File: toolchain/base/timings.h (hunk 2 above)
Comment (danakj) [reply toolchain/base/timings.h]: Oh yeah, done.

**Pull request merged**